### PR TITLE
Enable user defined cost dict for templates

### LINKED ISSUE #5232: Quantum cost of template substitution
### What is the expected enhancement?
Template substitution allows users to identify sub-circuits in a quantum circuit and replace them with equivalent quantum circuits with a lower implementation cost. For example `CNOT(0,1) CNOT(0,1)` is equivalent to the identity `I`. Since the implementation cost of `I` is lower than two CNOT gates this replacement makes sense. 

Currently, the implementation cost of a quantum circuit is evaluated in `TemplateSubstitution._quantum_cost(self, left, right)` in `template_substitution.py`. `_quantum_cost` evaluates the cost using a hard coded `cost_dict` instance. The user should be able to add/remove/modify elements from this dictionary as it is architecture dependent and the user may also whish to add his own gates to it.

`cost_dict` should therefore be an attribute of the `TemplateSubstitution` class that users can access when creating template matching transpiler passes.

### PR DESCRIPTION
**Pull request by eggerdj**
### Summary
Fixes #5232 by allowing the user to configure the cost dictionary of the template substitution.

### Details and comments
The user can provide an optional `dict` to `TemplateOptimization` which allows it to handle user defined gates and override the default costs if need be.

### CODE REVIEW

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 1]
```diff
@@ -62,6 +64,16 @@ def __init__(self, max_matches, circuit_dag_dep, template_dag_dep):
         self.unmatched_list = []
         self.dag_dep_optimized = DAGDependency()
         self.dag_optimized = DAGCircuit()
+        self.cost_dict = {'id': 0, 'x': 1, 'y': 1, 'z': 1, 'h': 1, 't': 1, 'tdg': 1, 's': 1,
```
Comment (ewinston) [inline qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:67]: These costs seem to be based on default decompositions defined for the gate. Occasionally the default decompositions change or a different target basis is used. I wonder if there is a way to make this initialization more aware of this?

Comment (eggerdj) [reply qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:67]: @ewinston I think there might be a way to do this using the `SessionEquivalenceLibrary` in Qiskit. One could e.g. do:

```
SessionEquivalenceLibrary = EquivalenceLibrary(base=StandardEquivalenceLibrary)
```
then this
```
iswap = SessionEquivalenceLibrary.get_entry(iSwapGate())[0]
```
gives you the decomposition of iswap into CNOT gates (or whatever mapping you have defined in the equivalence library). Then we could use these circuits to arrive at the `cost_dict` by counting the number of gates. Thoughts?

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 2]
```diff
@@ -63,6 +65,16 @@ def __init__(self, max_matches, circuit_dag_dep, template_dag_dep):
         self.dag_dep_optimized = DAGDependency()
         self.dag_optimized = DAGCircuit()
 
+        if user_cost_dict is not None:
+            self.cost_dict = dict(user_cost_dict)
```
Comment (lcapelluto) [inline qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:69]: This is a shallow copy?

**Review by ewinston — APPROVED** [review]

**Review by lcapelluto — APPROVED** [review]

File: releasenotes/notes/issue-5232-995e0d1a57c7187e.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
```
Comment (lcapelluto) [inline releasenotes/notes/issue-5232-995e0d1a57c7187e.yaml:2]: should it be `features`?

**Pull request merged**